File: packages/shared/src/physics.ts
import type { ShipState, ShipInput, ShipConfig } from './types';

/**
 * Apply rotation to ship based on left/right input.
 * Orientation is 0-1 (fraction of full rotation), 0 = east, CW positive.
 */
export function applyRotation(
  state: ShipState,
  input: ShipInput,
  config: ShipConfig,
  dt: number,
): void {
  if (input.left) {
    state.orientation -= config.rotation * dt;
  }
  if (input.right) {
    state.orientation += config.rotation * dt;
  }
  // Wrap to [0, 1)
  if (state.orientation < 0) {
    state.orientation += 1;
  }
  if (state.orientation >= 1) {
    state.orientation -= 1;
  }
}

/**
 * Apply thrust in facing direction. Does NOT apply any drag or decay.
 * When thrust is not pressed, velocity remains EXACTLY unchanged.
 */
export function applyThrust(
  state: ShipState,
  input: ShipInput,
  config: ShipConfig,
  dt: number,
): void {
  if (!input.thrust && !input.reverse) {
    return;
  }

  const thrustValue =
    input.afterburner && state.energy > 0 ? config.maxThrust : config.thrust;
  const angle = state.orientation * Math.PI * 2;
  const direction = input.reverse ? -1 : 1;
  state.vx += Math.cos(angle) * thrustValue * dt * direction;
  state.vy += Math.sin(angle) * thrustValue * dt * direction;
}

/**
 * Hard-clamp speed using Truncate method (NOT drag).
 * If speed exceeds limit, scale velocity to limit. Never reduces speed below limit.
 */
export function clampSpeed(
  state: ShipState,
  input: ShipInput,
  config: ShipConfig,
): void {
  const speedLimit =
    input.afterburner && state.energy > 0 ? config.maxSpeed : config.speed;
  const currentSpeed = Math.sqrt(state.vx * state.vx + state.vy * state.vy);

  if (currentSpeed > speedLimit) {
    const scale = speedLimit / currentSpeed;
    state.vx *= scale;
    state.vy *= scale;
  }
}

/**
 * Update energy: afterburner drain and recharge.
 */
export function updateEnergy(
  state: ShipState,
  input: ShipInput,
  config: ShipConfig,
  dt: number,
): void {
  if (input.afterburner && state.energy > 0) {
    state.energy -= config.afterburnerCost * dt;
  }
  state.energy += config.recharge * dt;
  state.energy = Math.max(0, Math.min(config.maxEnergy, state.energy));
}

/**
 * Full physics update for one tick. Order: thrust -> rotation -> speed clamp -> energy.
 * Position update and collision are handled separately.
 */
export function updateShipPhysics(
  state: ShipState,
  input: ShipInput,
  config: ShipConfig,
  dt: number,
): void {
  applyThrust(state, input, config, dt);
  applyRotation(state, input, config, dt);
  clampSpeed(state, input, config);
  updateEnergy(state, input, config, dt);
}

File: packages/server-rs/src/game/physics.rs
use std::f32::consts::TAU;

use crate::config::ShipConfig;

/// Ship state (position, velocity, orientation, energy).
#[derive(Debug, Clone)]
pub struct ShipState {
    pub x: f32,
    pub y: f32,
    pub vx: f32,
    pub vy: f32,
    pub orientation: f32, // 0-1 (fraction of full rotation, 0=east, CW positive)
    pub energy: f32,
}

/// Ship input flags.
#[derive(Debug, Clone, Default)]
pub struct ShipInput {
    pub left: bool,
    pub right: bool,
    pub thrust: bool,
    pub reverse: bool,
    pub afterburner: bool,
    pub multifire: bool,
}

/// Full physics update for one tick. Order: thrust -> rotation -> speed clamp -> energy.
pub fn update_ship_physics(state: &mut ShipState, input: &ShipInput, config: &ShipConfig, dt: f32) {
    apply_thrust(state, input, config, dt);
    apply_rotation(state, input, config, dt);
    clamp_speed(state, input, config);
    update_energy(state, input, config, dt);
}

fn apply_rotation(state: &mut ShipState, input: &ShipInput, config: &ShipConfig, dt: f32) {
    if input.left {
        state.orientation -= config.rotation * dt;
    }
    if input.right {
        state.orientation += config.rotation * dt;
    }
    if state.orientation < 0.0 {
        state.orientation += 1.0;
    }
    if state.orientation >= 1.0 {
        state.orientation -= 1.0;
    }
}

fn apply_thrust(state: &mut ShipState, input: &ShipInput, config: &ShipConfig, dt: f32) {
    if !input.thrust && !input.reverse {
        return;
    }

    let thrust_value = if input.afterburner && state.energy > 0.0 {
        config.max_thrust
    } else {
        config.thrust
    };
    let angle = state.orientation * TAU;
    let direction = if input.reverse { -1.0 } else { 1.0 };
    state.vx += angle.cos() * thrust_value * dt * direction;
    state.vy += angle.sin() * thrust_value * dt * direction;
}

fn clamp_speed(state: &mut ShipState, input: &ShipInput, config: &ShipConfig) {
    let speed_limit = if input.afterburner && state.energy > 0.0 {
        config.max_speed
    } else {
        config.speed
    };
    let current_speed = (state.vx * state.vx + state.vy * state.vy).sqrt();
    if current_speed > speed_limit {
        let scale = speed_limit / current_speed;
        state.vx *= scale;
        state.vy *= scale;
    }
}

fn update_energy(state: &mut ShipState, input: &ShipInput, config: &ShipConfig, dt: f32) {
    if input.afterburner && (input.thrust || input.reverse) && state.energy > 0.0 {
        state.energy -= config.afterburner_cost * dt;
    }
    state.energy += config.recharge * dt;
    state.energy = state.energy.clamp(0.0, config.max_energy);
}
